Partial date condition

joblo2213 opened the pull request:
A partialdate condition as discussed in #651.

It took some time to create this pr because I had some internet problems in the last week and also was bussy with other things...

> ## Partial date: `partialdate`
> 
> The current date must match the given pattern. You can specify the day of the month, the month or the year it must be that this condition returns true or combine them. You can also specify multiple days/months/years by just separating them by `,` or a interval by using `-`. If you have trouble understanding how this works have a look at the example.
> 
> The example is true between the 1st and the 5th or on the 20th of each month, but only in the year 2017.
> 
> **Example:** `partialdate day:1-5,20 year:2017`

Co0sh commented on src/main/java/pl/betoncraft/betonquest/conditions/PartialDateCondition.java:
```diff
@@ -0,0 +1,136 @@
+package pl.betoncraft.betonquest.conditions;
+
+import com.google.common.base.Joiner;
```
This import is unused.

Co0sh commented on src/main/java/pl/betoncraft/betonquest/conditions/PartialDateCondition.java:
```diff
@@ -0,0 +1,136 @@
+        }
+        if (!month.isEmpty()) {
+            if (!month.contains(current.get(Calendar.MONTH) + 1)) return false;
+            //Dont ask why +1: java.util.times is a complete mess (january is 0, december is 11,...)
```
It's a perfectly valid solution though.

Co0sh commented on src/main/java/pl/betoncraft/betonquest/conditions/PartialDateCondition.java:
```diff
@@ -0,0 +1,136 @@
+            }
+        }
+        String monthString = instruction.getOptional("month");
+        if (monthString != null) {
```
Lots of duplicated code. It works, but it's not clean. If you don't want to change it then it's fine. I'll just redirect eventual bug reports to you 😉 

In case you'd like to change it, here's my suggestion, the way I see it:
Create a `Range` class. It will take `min` and `max` values when created, and then will have `boolean checkNumber(int number)` method which will test if a number is in range. Then store those `Range`s in those maps above (`dayOfMonth`, `month`, `year`). Note that a single day could be represented as a `Range` with `min == max`.
You could check those ranges in the `check()` method with something like this:

```java
year.stream().anyMatch(range -> range.checkNumber(number));
```

Creating `Range`s could be easy - since each range is separated with a comma, you can split the string and iterate over those substrings, creating new `Range`s. Then, in the constructor, either parse the two numbers separated with dash or simply double a single number to both `min` and `max` fields. Additionally specify in the constructor lower and upper boundaries (like 1 and 31, or 1 and 12) and check the parsed `min` and `max` fields against them.

This way you have a nice, reusable `Range` which can be applied to all three time types (and more, you could even use it for hours, minutes or seconds; even weeks and days of weeks could work that way).

joblo2213 replied:
Yes IntellIJ also told me I should remove this duplicated code when I was working on this but back then I was to lazy. I'll try to have a look into it tomorrow evening.

Co0sh commented on src/main/java/pl/betoncraft/betonquest/conditions/PartialDateCondition.java:
```diff
@@ -0,0 +1,136 @@
+                        if (beginning >= end) throw new InstructionParseException("year " + beginning + " is before " + end);
+                        for (int i = beginning; i <= end; i++) {
+                            if (i < 1) throw new InstructionParseException(yearString + " contains invalid years");
+                            year.add(i);
```
User tries to be funny and types `1-2147483647`, server crashes.

joblo2213 replied:
💀 
Actually havn't thought about that as I thoght people who are smart enogh to write BetonQuest-Quest are also smart enogh not to try such expreiments.

But you are totally right we also have to write the plugin for Dan (a german saying, the long form would be "**D**ümmster **a**nzunehmender **N**utzer" which means as much as much as "the dumbest user to be assumed" 😄)

Namnodorel replied:
"as much as much"? :D

joblo2213 commented:
Just reworked the condition based on the suggestions of @Co0sh.
I'll go to get some sleep right now (tomorrow my alarm will ring at 5am 😔)
and add some comments and do some testing tomorrow evening.
I'll let you know when it is ready to merge.

joblo2213 commented:
Tested, fixed a bug (only checking first optional if multiple ones are used) and added some comments...
With the use of the TImeInterval we also don't need any cap limitting the size of the interval.
@Co0sh I think it's ready to be merged

Co0sh commented:
I'll be doing PRs soon, I just have a lot going on lately. Sorry guys 😐
